### PR TITLE
Generalization for ElementsAttr coverage

### PR DESCRIPTION
**Pull request by saienduri**
This PR addresses the issue that there was code written in IREE from a time when there was only DenseElementsAttr. DenseElementsAttr is a specific type and ElementsAttr the interface. There are actually multiple implementations of the interface now and DenseElementsAttr is being phased out for general usage. This PR deals with generalization, so we are not boxed into only supporting DenseElementsAttr.

### CODE REVIEW

**Comment by hanhanW** [comment]
It looks like you are replacing `DenseElementsAttr` with `ElementsAttr`. Can you elaborate why are we doing this?

**Comment by saienduri** [comment]
> It looks like you are replacing `DenseElementsAttr` with `ElementsAttr`. Can you elaborate why are we doing this?

Sure, the issue is that there was code written in IREE from a time when there was only DenseElementsAttr. Then DenseElementsAttr was partially generalized by making it expose the ElementsAttr interface but code wasn't updated. Now that DenseResourceElementsAttr is on the scene, we have to properly support both, at least in critical code paths or face crashes/failures/etc. This PR deals with generalization, so we are not boxed into only supporting DenseElementsAttr.

**Comment by stellaraccident** [comment]
> It looks like you are replacing `DenseElementsAttr` with `ElementsAttr`. Can you elaborate why are we doing this?

`DenseElementsAttr` is a specific type and `ElementsAttr` the interface. There are actually multiple implementations of the interface and `DenseElementsAttr` is being phased out for general usage.

File: compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalg.cpp [hunk 1]
```diff
@@ -1417,16 +1417,23 @@ struct ConstConverterTensor final
...
+                                                     replacementAttr);
+      return success();
+    } else {
+      auto denseAttr = dyn_cast<DenseElementsAttr>(constOp.getValue());
+      if (!denseAttr) {
+        return rewriter.notifyMatchFailure(
+            constOp,
+            "DenseElementsAttr cast failed (only DenseElementsAttr supported)");
       }
+      replacementAttr = denseAttr.mapValues(replacementType.getElementType(),
```
Comment (stellaraccident) [inline compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalg.cpp:1434]: We are going to need a more generic utility for mapping elements regardless of the backing type. Can you add a TODO (and reference an issue) suggesting that this be supported.

Comment (stellaraccident) [inline compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalg.cpp:1434]: Can you preserve the "Signedness conversion" comment from the original here? That is actually a really important hint to the reader.

**Review by stellaraccident — APPROVED** [review]
Couple of comment nits - otherwise, lgtm.

**Comment by hanhanW** [comment]
Thanks for the details, can you add them to PR description? That will help others to understand what's happening in the PR/commit better.

File: compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalg.cpp (hunk 1 above)
Comment (saienduri) [reply compiler/src/iree/compiler/InputConversion/StableHLO/StableHLOToLinalg.cpp:1434]: created issue and marked as TODO here

**Pull request merged**